cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java

```diff
@@ -0,0 +1,522 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * Kubernetes version of a TwillRunner.
+ *
+ * Runs a program in Kubernetes by creating a config-map of the app spec and program options resources that are
+ * expected to be found in the local files of the RuntimeSpecification for the TwillRunnable.
+ * A deployment is created that mounts the created config-map.
+ *
+ * Most of these operations are no-ops as many of these methods and pretty closely coupled to the Hadoop implementation
+ * and have no analogy in Kubernetes.
+ */
+class KubeTwillPreparer implements TwillPreparer {
+  private static final Logger LOG = LoggerFactory.getLogger(KubeTwillPreparer.class);
+  // These are equal to the constants used in DistributedProgramRunner
+  private static final String APP_SPEC = "appSpec.json";
+  private static final String PROGRAM_OPTIONS = "program.options.json";
+  private final ApiClient apiClient;
+  private final String kubeNamespace;
+  private final String image;
+  private final PodInfo podInfo;
+  private final RunId runId;
+  private final List<URI> resources;
+  private final Set<String> runnables;
+  private final Map<String, Map<String, String>> environments;
+  private final V1ObjectMeta resourceMeta;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final URI appSpec;
+  private final URI programOptions;
+  private final int memoryMB;
+  private final int vcores;
+
+  KubeTwillPreparer(ApiClient apiClient, String kubeNamespace, String image, PodInfo podInfo, RunId runId,
+                    TwillSpecification spec, V1ObjectMeta resourceMeta, DiscoveryServiceClient discoveryServiceClient) {
+    // only expect one runnable for now
+    if (spec.getRunnables().size() != 1) {
+      throw new IllegalStateException("Kubernetes runner currently only supports one Twill Runnable");
+    }
+    this.apiClient = apiClient;
+    this.kubeNamespace = kubeNamespace;
+    this.image = image;
+    this.runId = runId;
+    this.podInfo = podInfo;
+    this.runnables = spec.getRunnables().keySet();
+    this.resources = new ArrayList<>();
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.environments = new HashMap<>();
+    for (String runnable : runnables) {
+      environments.put(runnable, new HashMap<>());
+    }
+    this.resourceMeta = resourceMeta;
+    RuntimeSpecification runtimeSpecification = spec.getRunnables().values().iterator().next();
+    ResourceSpecification resourceSpecification = runtimeSpecification.getResourceSpecification();
+    appSpec = runtimeSpecification.getLocalFiles().stream()
```

Comment:
To be consistent with other fields in the constructor, should be `this.appSpec`.

Reply:
done